gap/digraph.gi

```diff
@@ -1368,6 +1423,259 @@ function(filt, n, p)
+
+            # Check that the edge doesn't already exist
+            if (not (i in adjacencyList[circuitVertex])) then
+                # First we guarantee that we get a connected graph by checking if the vertex isn't already connected
+                if (not (i in connectedVertices)) then
+                    Add(adjacencyList[circuitVertex], i);
+                    Add(connectedVertices, i);
+                    circuitVertex := i;
+                    break;
+                elif Float(Random([0 .. 99])/100) < Float(p) then
+                    Add(adjacencyList[circuitVertex], i);
+                    circuitVertex := i;
+                    break;
+                fi;
+            fi;
+        od;
+
+        # If there are not more vertices to consider, we can end the circuit
+        if Length(verticesToConsider[circuitVertex]) = 0 then
+            continueCircuit := false;
+        fi;
+    od;
+
+    # Finish the circuit by adding an edge back to the start vertex (if it isn't already at the start vertex)
+    if circuitVertex <> startVertex then
+        # If an edge already exists between the finish vertex and the start vertex,
+        # we must find another path to the start vertex to avoid generating a multidigraph.
+        while (startVertex in adjacencyList[circuitVertex]) do
+            # Here we consider all possible edges again, including previously rejected edges, to guarantee a path
+            # back to the start
+            verticesToCheck := [1 .. n];
+            while Length(verticesToCheck) > 0 do
+                i := Random(verticesToCheck);
+                Remove(verticesToCheck, Position(verticesToCheck, i));
+                if (not (i in adjacencyList[circuitVertex])) then
+                    Add(adjacencyList[circuitVertex], i);
+                    circuitVertex := i;
+                    break;
+                fi;
+            od;
+        od;
+        Add(adjacencyList[circuitVertex], startVertex);
+    fi;
+
+    graph := Digraph(adjacencyList);
+    return graph;
+end);
+
+InstallMethod(RandomDigraphCons,
+"for IsConnectedDigraph, a positive integer, and a float",
+[IsConnectedDigraph, IsPosInt, IsFloat],
+function(filt, n, p)
+    local adjacencyList, vertices, startVertex, tree, x, i, j, graph;
+
+    adjacencyList := [];
+
+    for i in [1 .. n] do
+        Add(adjacencyList, []);
+    od;
+
+    vertices:= [1 .. n];
+
+    # Starting from a random vertex, we first create a tree to guarantee connectivity
+    startVertex := Remove(vertices, Random(vertices));
+    tree := [startVertex];
+
+    # While there are n-1 remaining vertices to be added to the tree
+    for x in [1 .. n-1] do
+        # Create an edge from a random vertex in the tree, to a random vertex outside of it
+        i := Random(tree);
+        j := Random(vertices);
+        Remove(vertices, Position(vertices, j));
+        Add(tree, j);
+        Add(adjacencyList[i], j);
+    od;
```

Comment:
James was suggesting that this should be a separate function for generating random trees.  But for now I suggest we leave it here, and remember for later.